lib/maths/analytics/unittest/CBoostedTreeTest.cc

```diff
@@ -3085,7 +3085,8 @@ BOOST_AUTO_TEST_CASE(testEstimateMemory) {
         LOG_DEBUG(<< "estimated memory usage = " << estimatedMemory);
         LOG_DEBUG(<< "high water mark = " << predictInstrumentation.maxMemoryUsage());
 
-        BOOST_TEST_REQUIRE(predictInstrumentation.maxMemoryUsage() <= estimatedMemory);
+        BOOST_TEST_REQUIRE(predictInstrumentation.maxMemoryUsage() <=
+                           (estimatedMemory * 101) / 100);
```

Comment:
Instead of adding the uplift here it should be added in https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L208-L210 https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L220-L221  https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L492-L493 and https://github.com/elastic/ml-cpp/blob/c28c82c0c29e13990d169fbea5a6e1836e3a1b91/include/core/CMemoryDef.h#L529-L530

The new memory layout is detailed here: https://www.boost.org/doc/libs/1_83_0/libs/unordered/doc/html/unordered.html#structures_closed_addressing_containers

The formulas should be updated to match this. As far as I can see the key part is:

> there’s only 4 bits of space overhead per bucket introduced by the bucket groups.

So we should add on an extra half a byte per bucket (rounded up to the next higher multiple of `sizeof(std::size_t))`) to account for these bucket groups.

Please give this a try and see if it allows the 1% hack here to be removed. If it does then please also add a link to that section of the Boost docs as part of a comment to explain why we're adding this strange amount.